Update Azure Blob to 12.1.0

maganaluis opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kubeflow/kfserving/blob/master/docs/DEVELOPER_GUIDE.md
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://github.com/kubeflow/kfserving/issues/923

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:
```
1. Upgraded Azure Blob to 12.1.0 as part of the storage initializer in KFServing. 
```

yuzisun commented:
@maganaluis Thanks for taking care of this issue! Can you check the CLA seems like the commit includes some other author.

maganaluis commented:
@yuzisun CLA should be good now.

yuzisun commented on python/kfserving/kfserving/storage.py:
```diff
@@ -173,7 +172,9 @@ def _download_blob(uri, out_dir: str):  # pylint: disable=too-many-locals
 
             logging.info("Downloading: %s to %s", blob.name, dest_path)
-            block_blob_service.get_blob_to_path(container_name, blob.name, dest_path)
+            downloader = blob_service_client.get_container_client(container_name).download_blob(blob.name)
```
`get_container_client` gets called twice here.

maganaluis replied:
Yup, will take care of this. Let's continue the discussion on how to allow users to setup the AZURE_STORAGE_CONNECTION_STRING, as this will slightly broaden the scope of this PR.

yuzisun commented on python/kfserving/kfserving/storage.py:
```diff
@@ -150,15 +149,15 @@ def _download_blob(uri, out_dir: str):  # pylint: disable=too-many-locals
-            block_blob_service = BlockBlobService(account_name=account_name)
-            blobs = block_blob_service.list_blobs(container_name, prefix=prefix)
-        except Exception:  # pylint: disable=broad-except
+            connect_str = os.getenv('AZURE_STORAGE_CONNECTION_STRING')
```
how is this environment variable going to be set?

maganaluis replied:
Good question. The connection string, is highlighted in the documentation, for best practices.

https://docs.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-python

However the storage initializer sidecar will not have a way to provide this information or a way to support it. Do we want to add support here? I think that would be the best approach. 

https://github.com/kubeflow/kfserving/blob/master/docs/samples/storage/azure/README.md
https://github.com/kubeflow/kfserving/blob/42007c532286a1e43893ef2be03b15e104bfd7a4/pkg/credentials/azure/azure_secret.go

Let me know your thoughts.

yuzisun commented:
Thanks @maganaluis ! would be highly appreciable if you can help add an e2e test for azure.

yuzisun commented:
@maganaluis can you help resolve the conflicts here?

maganaluis commented:
I corrected the conflicts. Could you check this https://github.com/kubeflow/kfserving/issues/1671 we are working on an event driven architecture with KFServing :)

maganaluis commented:
@yuzisun I corrected the conflicts. Unrelated, but could you provide input on https://github.com/kubeflow/kfserving/issues/1671?  :)

yuzisun commented:
> @yuzisun I corrected the conflicts. Unrelated, but could you provide input on #1671? :)
> 
> Also note I removed checking for env variables, given this code runs within the storage sidecar there is currently no way to set this up.

@maganaluis You should be able to easily test storage initializer with `make deploy-dev-storageInitializer`! Can you also help fix the lint issue?

yuzisun commented:
/ok-to-test

yuzisun commented:
@maganaluis Can you help fix the lint and tests?

maganaluis commented:
@yuzisun I corrected the test cases, there is a lot of lint errors not related to my change. Is there a particular way to run the lint check? 

Also, I switched to use azure-identity to create the token. This seems to be the best way to authenticate with the latest version of Azure Blob. I tested this on a minikube cluster with the latest version of KFServing.

yuzisun commented:
> @yuzisun I corrected the test cases, there is a lot of lint errors not related to my change. Is there a particular way to run the lint check?

./python/kfserving/kfserving/storage.py:27:16: W291 trailing whitespace
./python/kfserving/test/test_azure_storage.py:48:1: E302 expected 2 blank lines, found 1

@maganaluis there are two lint errors which should be an easy fix, thanks.

maganaluis commented:
I pushed some more changes, after the previous commit some test cases had to change.

yuzisun commented:
@maganaluis this looks good to me, are we able to add an e2e test for this? I am not sure how to verify this change.

maganaluis commented:
> @maganaluis this looks good to me, are we able to add an e2e test for this? I am not sure how to verify this change.

I can add a e2e test case, but I don't see e2e test cases for the other storage options, you would have to have a storage account for each cloud provider.

yuzisun commented on python/kfserving/kfserving/storage.py:
```diff
@@ -168,16 +171,13 @@ def _download_blob(uri, out_dir: str):  # pylint: disable=too-many-locals
+        token = Storage._get_azure_storage_token()
+        if token is None:
+            logging.warning("Azure credentials not found, retrying anonymous access")
+        blob_service_client = BlobServiceClient(account_url, credential=token)
```
@maganaluis if the token is None, this should still work right? Does azure support public storage blobs?

yuzisun replied:
oh I found the doc https://docs.microsoft.com/en-us/python/api/overview/azure/storage-blob-readme?view=azure-python#types-of-credentials

maganaluis replied:
Correct, however that's unlikely to be used. I did verify that Azure AD works. The current documentation still holds well.

yuzisun commented on python/kfserving/kfserving/storage.py:
```diff
@@ -192,7 +192,10 @@ def _download_blob(uri, out_dir: str):  # pylint: disable=too-many-locals
 
             logging.info("Downloading: %s to %s", blob.name, dest_path)
-            block_blob_service.get_blob_to_path(container_name, blob.name, dest_path)
+            Path(os.path.dirname(dest_path)).mkdir(parents=True, exist_ok=True)
```
With this do we still need the above code from 183 to 192 to create the directories?

maganaluis replied:
That's right, they were unnecessary, I removed those lines.

yuzisun commented:
Thanks @maganaluis for your awesome work!

/lgtm
/approve
